lib/kredis/callbacks_proxy.rb

```diff
@@ -22,7 +22,7 @@ def method_missing(method, *args, **kwargs, &block)
 
     if CALLBACK_OPERATIONS[@type.class]&.include? method
       if @callback.respond_to? :call
-        @callback.call(@record)
+        @callback.call(@record || @type)
```

Comment:
I know this looks like a straight shot but it makes it hard to reason about which paths this if/elsif are effected by.

Can we extract the `@record` requirement to just the attributes callsite and have it enrich the provided original callback with the record access?

Reply:
I'm afraid I don't quite follow. Could you use more/different words please?

Reply:
Sure, what I'm trying to get at is that the `CallbacksProxy`'s original contract is now being stretched into this new context — one where there isn't a record. I think we should follow that through by yanking the record state out of `CallbacksProxy` and use function composition to capture it instead and do something like:

```ruby
def kredis_connection_with(method, name, key, **options)
  ivar_symbol = :"@#{name}_#{method}"
  type = method.to_s.sub("kredis_", "")
  after_change = options.delete(:after_change)

  define_method(name) do
    if instance_variable_defined?(ivar_symbol)
      instance_variable_get(ivar_symbol)
    else
      new_type = Kredis.send(type, kredis_key_evaluated(key) || kredis_key_for_attribute(name), **options)

      instance_variable_set ivar_symbol,
        after_change ? enrich_after_change_with_record_access(new_type, after_change) : new_type
    end
  end
end

# private method defined at the instance level:
def enrich_after_change_with_record_access(type, original_after_change)
  # May need a record = self here in case the scope capture can't rely on self.
  case original_after_change
  when Proc   then Kredis::CallbacksProxy.new(type, ->(_) { original_after_change.call(self) })
  when Symbol then Kredis::CallbacksProxy.new(type, ->(_) { send(original_after_change) })
  end
end
```

Note: this change also opts out of the proxy if there's no `after_change` such that we don't pay the `method_missing` cost needlessly.

Reply:
Ah, that makes a lot of sense now. Modified!